### PR TITLE
[SOAR-16699] [Carbon Black Cloud] Monitor alerts task implementation

### PR DESCRIPTION
**Pull request by joneill-r7**
## Proposed Changes
[[C2C] Carbon Black task implementation](https://rapid7.atlassian.net/browse/SOAR-16699)

### Description
I've tried to keep the commits in a logical order for anyone who is reviewing. I've yet to add unit tests which will be coming shortly but opening this to start getting reviews. 

  - Small changes to bump to the latest SDK and update the plugin spec so that we keep the version history etc each time that we perform a refresh. 
  - Updated the existing actions to use a generic `request_to_api` now that it has built upon to use a get request for the observation job ID for the task. 
  - A few new utils (hashing, request error handling) and rate limiting exception added. 
  - Task implementation which has been based from the following wiki - [Carbon Black Research](https://rapid7.atlassian.net/wiki/spaces/KOMAND/pages/677216868/Carbon+Black+Research)
    - The task is to follow our normal task workflow of reaching out to 3rd party endpoints and polling data during a given timeframe. In this instance we need to trigger an async job (for observations) and then query for completion. 
    - CB rate limits due to the number of requests in 5 minute period - they are to return a `Retry-After` header when this is imposed but I've been unable to hit this during my testing so i've had to add a logger to see what this value is given in ( standards say it can be seconds or a timestamp), and added a standard 5 minute delay. 
    - Due to the page sizing (2500) available in one request, I've went for the approach to iterate forwards and backwards (`_dedupe_and_get_last_time` over the results to dedupe alerts and get next timestamp for the next run. 

## PR Requirements

Developers, verify you have completed the following items by checking them off:

### Testing
Manual testing of the following scenarios with logs -

- First successful run -
```
Plugin task beginning execution...
rapid7/VMware Carbon Black Cloud:2.2.0. Step name: monitor_alerts
No last_alert_time within state. No last_observation_time within state. Applying the following start times: alerts='2024-04-25T08:44:52.620258Z' and observations='2024-04-25T08:44:52.620258Z'
No observation job ID found in state, triggering a new job...
Triggering observation search using parameters {'start': '2024-04-25T08:44:52.620258Z', 'end': '2024-04-25T08:49:52.620258Z'}
Saving observation job ID 5d0ea65a-d7e2-4b66-a544-c12dda7e9ef3-sqs to the state. Will query this after polling for alerts...
No last_observation_time in the state, saving checkpoint as 2024-04-25T08:44:52.620258Z
Querying alerts using parameters {'start': '2024-04-25T08:44:52.620258Z', 'end': '2024-04-25T08:49:52.620258Z'}
Expecting to dedupe 0 based on previously stored hashes.
Received 1, and after dedupe there is 1 results.
Setting last_alert_time to last parsed time: '2024-04-25T08:45:45.532Z'
last_alert_time set to: 2024-04-25T08:45:45.532Z, has_more_pages=False
Get observation results from saved ID: 5d0ea65a-d7e2-4b66-a544-c12dda7e9ef3-sqs
Expecting to dedupe 0 based on previously stored hashes.
Received 71, and after dedupe there is 71 results.
Setting last_observation_time to last parsed time: '2024-04-25T08:49:44.498Z'
last_observation_time set to: 2024-04-25T08:49:44.498Z, has_more_pages=False
Returning a combined total of 72 alerts and observations, with has_more_pages=False
Plugin task finished execution...
```
- Second run using previous state -
```
Plugin task beginning execution...
rapid7/VMware Carbon Black Cloud:2.2.0. Step name: monitor_alerts
Applying the following start times: alerts='2024-04-25T08:45:45.532Z' and observations='2024-04-25T08:49:44.498Z'
No observation job ID found in state, triggering a new job...
Triggering observation search using parameters {'start': '2024-04-25T08:49:44.498Z', 'end': '2024-04-25T08:50:42.369435Z'}
Saving observation job ID aab7b08f-0a80-4438-ba87-2631ba0b6925-sqs to the state. Will query this after polling for alerts...
Querying alerts using parameters {'start': '2024-04-25T08:45:45.532Z', 'end': '2024-04-25T08:50:42.369435Z'}
Expecting to dedupe 1 based on previously stored hashes.
Received 1, and after dedupe there is 0 results.
last_alert_time set to: 2024-04-25T08:45:45.532Z, has_more_pages=False
Get observation results from saved ID: aab7b08f-0a80-4438-ba87-2631ba0b6925-sqs
Expecting to dedupe 1 based on previously stored hashes.
Received 16, and after dedupe there is 16 results.
Setting last_observation_time to last parsed time: '2024-04-25T08:50:36.834Z'
last_observation_time set to: 2024-04-25T08:50:36.834Z, has_more_pages=False
Returning a combined total of 16 alerts and observations, with has_more_pages=False
Plugin task finished execution...
```
- Second run with rate limiting value imposed -
```
Plugin task beginning execution...
rapid7/VMware Carbon Black Cloud:2.2.0. Step name: monitor_alerts
Rate limit value stored in state: 2024-04-25T13:34:03.626Z. Still within rate limiting period, skipping task execution...
Plugin task finished execution...
```
- Second run with rate limiting value now expired - 
```
Plugin task beginning execution...
rapid7/VMware Carbon Black Cloud:2.2.0. Step name: monitor_alerts
Rate limit value stored in state: 2024-04-24T13:34:03.626Z. However no longer in rate limiting period, so task can be executed...
Applying the following start times: alerts='2024-04-24T13:33:32.358Z' and observations='2024-04-24T13:34:03.626Z'
No observation job ID found in state, triggering a new job...
.....
```
- Pagination variable dropped to 1 to force `has_more_pages_true`
```
Plugin task beginning execution...
rapid7/VMware Carbon Black Cloud:2.2.0. Step name: monitor_alerts
Applying the following start times: alerts='2024-04-24T13:33:32.358Z' and observations='2024-04-24T13:34:03.626Z'
No observation job ID found in state, triggering a new job...
Triggering observation search using parameters {'start': '2024-04-24T13:34:03.626Z', 'end': '2024-04-24T15:33:15.275622Z'}
Saving observation job ID 4e45fc0d-6d11-4f59-b6dd-bd205725a528-sqs to the state. Will query this after polling for alerts...
Querying alerts using parameters {'start': '2024-04-24T13:33:32.358Z', 'end': '2024-04-24T15:33:15.275622Z'}
Have not got all alerts for given time period. Returning has more pages true...
Expecting to dedupe 1 based on previously stored hashes.
Received 1, and after dedupe there is 0 results.
last_alert_time set to: 2024-04-24T13:33:32.358Z, has_more_pages=True
Get observation results from saved ID: 4e45fc0d-6d11-4f59-b6dd-bd205725a528-sqs
Expecting to dedupe 1 based on previously stored hashes.
Received 1, and after dedupe there is 0 results.
more data is available on the API - setting has more pages to true...
last_observation_time set to: 2024-04-24T13:34:03.626Z, has_more_pages=True
Returning a combined total of 0 alerts and observations, with has_more_pages=True
Plugin task finished execution...
```

#### Unit Tests

Review our documentation on [generating](https://docs.rapid7.com/insightconnect/unit-test-generation) and [writing](https://docs.rapid7.com/insightconnect/unit-test-primer) plugin unit tests

- [ ] Unit tests written for any new or updated code <- outstanding
- [ ] Updated unit tests to now use request_to_api and updated mocks. 

#### In-Product Tests

If you are an InsightConnect customer or have access to an InsightConnect instance, the following in-product tests should be done:

- [ ] Screenshot of job output with the plugin changes
- [ ] Screenshot of the changed connection, actions, or triggers input within the InsightConnect workflow builder

### Style

Review the [style guide](https://docs.rapid7.com/insightconnect/style-guide/)

- [ ] For dependencies, pin [OS package](https://docs.rapid7.com/insightconnect/style-guide/#dockerfile) and [Python package](https://docs.rapid7.com/insightconnect/style-guide/#requirements.txt) versions
- [ ] For security, set least privileged account with ``USER nobody`` in the ``Dockerfile`` when possible
- [ ] For size, use the [slim SDK images](https://docs.rapid7.com/insightconnect/sdk-guide/#sdk-guide) when possible: ``rapid7/insightconnect-python-3-38-slim-plugin:{sdk-version-num}`` and ``rapid7/insightconnect-python-3-38-plugin:{sdk-version-num}``
- [ ] For error handling, use of [PluginException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations/#plugin-exceptions) and [ConnectionTestException](https://docs.rapid7.com/insightconnect/error-handling-in-integrations#connection-exceptions)
- [ ] For logging, use [self.logger](https://docs.rapid7.com/insightconnect/sdk-guide/#logging)
- [ ] For docs, use [changelog style](https://docs.rapid7.com/insightconnect/style-guide/#changelog)
- [ ] For docs, validate markdown with ``insight-plugin validate`` which calls ``icon_validate`` to lint ``help.md``

### Functional Checklist
- [ ] Work fully completed
- [ ] Functional
  - [ ] Any new actions/triggers include JSON [test files](https://docs.rapid7.com/insightconnect/style-guide/#tests) in the `tests/` directory created with `insight-plugin samples`
  - [ ] Tests should all pass unless it's a negative test. Negative tests have a naming convention of `tests/$action_bad.json`
  - [ ] Unsuccessful tests should fail by raising an exception causing the plugin to die and an object should be returned on successful test
  - [ ] Add functioning test results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -T tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run -T all --debug --jq` (use PR format at end)
  - [ ] Add functioning run results to PR, sanitize any output if necessary
    * Single action/trigger `insight-plugin run -R tests/example.json --debug --jq`
    * All actions/triggers shortcut `insight-plugin run --debug --jq` (use PR format at end)

### Assessment

You must validate your work to reviewers:

1. Run `insight-plugin validate` and make sure everything passes
2. Run the assessment tool: `insight-plugin run -A`. For single action validation: `insight-plugin run tests/{file}.json -A`
3. Copy (`insight-plugin ... | pbcopy`) and paste the output in **a new post** on this PR
4. Add required screenshots from the In-Product Tests section

### CODE REVIEW

File: plugins/carbon_black_cloud/icon_carbon_black_cloud/tasks/monitor_alerts/task.py [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+
+        if observation_job_id:
+            self.logger.info(
+                f"Saving observation job ID {observation_job_id} to the state. "
+                f"Will query this after polling for alerts..."
+            )
+            state[LAST_OBSERVATION_JOB] = observation_job_id
+
+            if not state.get(LAST_OBSERVATION_TIME):
+                # We should only hit this when we have *never* returned any observations, but after we've
```
Comment (dlaverty-r7) [inline plugins/carbon_black_cloud/icon_carbon_black_cloud/tasks/monitor_alerts/task.py:147]: Maybe reword this comment as its a little confusing?

File: plugins/carbon_black_cloud/icon_carbon_black_cloud/tasks/monitor_alerts/task.py [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+            description=Component.DESCRIPTION,
+            input=MonitorAlertsInput(),
+            output=MonitorAlertsOutput(),
+            state=MonitorAlertsState(),
+        )
+
+    def run(self, params={}, state={}, custom_config={}):  # pylint: disable=unused-argument # noqa: MC0001
+        alerts_and_observations, has_more_pages, observations_has_more_pages = [], False, False
+        try:
+            rate_limited = state.get(RATE_LIMITED)
```
[the remark this answers is not in the document]
Comment (joneill-r7) [reply plugins/carbon_black_cloud/icon_carbon_black_cloud/tasks/monitor_alerts/task.py:45]: Oops, no I haven't removed it. My bad in rushing to get the PR out. Will remove next commit!

File: plugins/carbon_black_cloud/icon_carbon_black_cloud/tasks/monitor_alerts/task.py (hunk 1 above)
Comment (joneill-r7) [reply plugins/carbon_black_cloud/icon_carbon_black_cloud/tasks/monitor_alerts/task.py:147]: Yeah totally agree, even I'm questioning what I meant by this 😆 will update!

File: plugins/carbon_black_cloud/icon_carbon_black_cloud/tasks/monitor_alerts/task.py [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+# State held values
+RATE_LIMITED = "rate_limited_until"
+LAST_ALERT_TIME = "last_alert_time"
+LAST_ALERT_HASHES = "last_alert_hashes"
+LAST_OBSERVATION_TIME = "last_observation_time"
+LAST_OBSERVATION_HASHES = "last_observation_hashes"
+LAST_OBSERVATION_JOB = "last_observation_job"
+
+# CB can return 10K per API and suggest that if more than this is returned to then query from last event time.
```
Comment (jerichardson-r7) [inline plugins/carbon_black_cloud/icon_carbon_black_cloud/tasks/monitor_alerts/task.py:25]: Appreciation for this comment 👏

**Review by jerichardson-r7 — APPROVED** [review]
This all looks great Johnny!! Really appreciate the very clear comments on everything as it made reading through the task code easier when this can usually be hard to understand with so much going on! Weird CB API only brings back 10 results for that one if you don't give it a row param, think its good you've set it to the same number as the other API!

**Review by igorski-r7 — APPROVED** [review]

File: plugins/carbon_black_cloud/icon_carbon_black_cloud/actions/get_agent_details/schema.py [hunk 4]
```diff
@@ -4,7 +4,7 @@
 
 
 class Component:
-    DESCRIPTION = "Get Agent Details"
+    DESCRIPTION = "Get agent details"
```
[the remark this answers is not in the document]
Comment (joneill-r7) [reply plugins/carbon_black_cloud/icon_carbon_black_cloud/actions/get_agent_details/schema.py:9]: Checked this and the get_agent_details has been updated. Tested the action and still works 👍

**Review by ekelly-r7 — APPROVED** [review]

**Review by dlaverty-r7 — APPROVED** [review]
Lgtm - one small query around job finishing.

File: plugins/carbon_black_cloud/icon_carbon_black_cloud/tasks/monitor_alerts/task.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        # Strange CB API behaviour, unless rows param is specified it only returns 10 results
+        url = f"{self.connection.base_url}/{endpoint}?rows={PAGE_SIZE}"
+        self.logger.info(f"Get observation results from saved ID: {job_id}")
+        observation_json = self.connection.request_api(
+            url,
+            request_method="GET",
+        )
+        if observation_json.get("contacted") != observation_json.get("completed"):
+            self.logger.info("Job is not yet finished running, will get results in next task execution...")
+            has_more_pages = True  # trigger again as it should be finished imminently (jobs run for a max of 3 minutes)
```
[the remark this answers is not in the document]
Comment (joneill-r7) [reply plugins/carbon_black_cloud/icon_carbon_black_cloud/tasks/monitor_alerts/task.py]: Had a read at the docs and they do state they can return an incomplete difference of 1 if the job is during a high activity period. I've included a 4 minute grace period to pull available results. 
> The job will be complete when contacted == completed in the response however during high usage a searcher may fail leaving a difference of 1. To prevent an infinite loop, ensure you add a timeout of 3 mins as a job's maximum active time is limited to 3 mins.

File: plugins/carbon_black_cloud/icon_carbon_black_cloud/tasks/monitor_alerts/task.py [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+        """
+        Jobs can only run within CB for a maximum of 3 minutes, allow a time frame of 4 minutes to complete otherwise we
+        then we can assume that the job has been cancelled due to high usage on their platform and there will be
+        a difference of 1 between the contacted and completed values and we should parse the available observations.
+        """
+        job_cut_off = (self._get_current_time() - timedelta(minutes=4)).strftime(TIME_FORMAT)
+
+        if job_cut_off > job_start_time:
+            self.logger.info(
+                f"Job has exceeded max run time. Started at {job_start_time}. Parsing available results..."
```
Comment (dlaverty-r7) [inline plugins/carbon_black_cloud/icon_carbon_black_cloud/tasks/monitor_alerts/task.py:293]: (Suggestion): Is there a job id or similar we could log here for troubleshooting?

**Review by dlaverty-r7 — APPROVED** [review]
Thanks for making the update around checking the job run time @joneill-r7 . Small suggestion around logging otherwise lgtm.

File: plugins/carbon_black_cloud/icon_carbon_black_cloud/tasks/monitor_alerts/task.py (hunk 6 above)
Comment (joneill-r7) [reply plugins/carbon_black_cloud/icon_carbon_black_cloud/tasks/monitor_alerts/task.py]: We look the job ID just before this, but added into this log line as well - 
```
{"message": "Get observation results from saved ID: example-sqs", "request_path": "/api/v1/tasks/monitor_alerts", "level": "info", "timestamp": "2024-05-01 13:36.06"}
{"message": "Job (example-sqs) has exceeded max run time. Started at 2024-04-01T13:34:03.626Z. Parsing available results...", "request_path": "/api/v1/tasks/monitor_alerts", "level": "info", "timestamp": "2024-05-01 13:36.08"}
{"message": "Expecting to dedupe a max of 1 based on previously stored hashes.", "request_path": "/api/v1/tasks/monitor_alerts", "level": "info", "timestamp": "2024-05-01 13:36.08"}
```

**Review by rbowden-r7 — APPROVED** [review]
